### PR TITLE
fix(pubsub): Pass headers for PubSub system so that users can get the properties 

### PR DESCRIPTION
**Pull request by XavierGeerinck**
Signed-off-by: Xavier Geerinck <xavier.geerinck@gmail.com>

# Description

## Issue reference

Please reference the issue this PR will close: #387

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [x] Code compiles correctly
- [x] Created/updated tests
- [ ] Extended the documentation

### CODE REVIEW

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md [hunk 1]
```diff
@@ -133,6 +133,8 @@ Subscribing to messages can be done in several ways to offer flexibility of rece
 - Direct susbcription with options through the `subscribeWithOptions` method
 - Subscription afterwards through the `susbcribeOnEvent` method
 
+Each time an event arrives, we then pass its body as `data` and the headers `headers` which can contain properties of the event publisher (e.g. a device id from IoT Hub)
```
Comment (shubham1172) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:136]:
```suggestion
Each time an event arrives, we pass its body as `data` and the headers as `headers`, which can contain properties of the event publisher (e.g., a device ID from IoT Hub)
```

File: src/implementation/Client/GRPCClient/invoker.ts [hunk 2]
```diff
@@ -47,7 +47,7 @@ export default class GRPCClientInvoker implements IClientInvoker {
       };
     }
 
-    if (method !== HttpMethod.GET && data !== {}) {
```
Comment (shubham1172) [inline src/implementation/Client/GRPCClient/invoker.ts:50]: Any specific reason for including an empty body in `fetchOptions`?

File: test/e2e/http/server.test.ts [hunk 3]
```diff
@@ -129,6 +131,23 @@ describe("http/server", () => {
...
 
+    it("should be able to receive events with their respective headers", async () => {
+      await server.client.pubsub.publish("pubsub-redis", "topic-5", { hello: "world" });
+
+      // Delay a bit for event to arrive
+      await new Promise((resolve, _reject) => setTimeout(resolve, 250));
+
+      expect(mockPubSubWithHeaders.mock.calls.length).toBe(1);
+
+      // Also test for receiving data
```
Comment (shubham1172) [inline test/e2e/http/server.test.ts:142]: Do you also want to check `mock.calls[0][0]` for data?

File: test/e2e/grpc/server.test.ts [hunk 4]
```diff
@@ -130,6 +132,19 @@ describe("grpc/server", () => {
...
 
+    it("should be able to receive events with their respective headers", async () => {
+      await server.client.pubsub.publish("pubsub-redis", "topic-5", { hello: "world" });
+
+      // Delay a bit for event to arrive
+      await new Promise((resolve, _reject) => setTimeout(resolve, 250));
+
+      expect(mockPubSubWithHeaders.mock.calls.length).toBe(1);
+
+      // Also test for receiving data
```
Comment (shubham1172) [inline test/e2e/grpc/server.test.ts:143]: Same here, to check for data.

File: src/implementation/Client/GRPCClient/invoker.ts (hunk 2 above)
Comment (XavierGeerinck) [reply src/implementation/Client/GRPCClient/invoker.ts:50]: It breaks typescript as this doesn't do anything in Javascript

**Review by shubham1172 — APPROVED** [review]
LGTM! 🥳 Just added some small comments and the build needs to be fixed.

File: test/e2e/http/server.test.ts (hunk 3 above)
Comment (XavierGeerinck) [reply test/e2e/http/server.test.ts:142]: I don't think so seeing that there are 3 checks under it so the test will fail in any case

File: test/e2e/grpc/server.test.ts (hunk 4 above)
Comment (XavierGeerinck) [reply test/e2e/grpc/server.test.ts:143]: I don't think so seeing that there are 3 checks under it so the test will fail in any case

**Review by shubham1172 — DISMISSED** [review]

File: test/e2e/http/server.test.ts [hunk 5]
```diff
@@ -23,6 +23,7 @@ describe("http/server", () => {
   let server: DaprServer;
   const mockBindingReceive = jest.fn(async (_data: object) => console.log("mockBindingReceive"));
   const mockPubSub = jest.fn(async (_data: object) => null);
+  const mockPubSubWithHeaders = jest.fn(async (_data: object, _headers: object) => null);
```
Comment (shubham1172) [inline test/e2e/http/server.test.ts:26]: Can you please add this under `beforeEach` as `mockPubSubWithHeaders.mockClear();` (both http and grpc)?

**Review by shubham1172 — APPROVED** [review]

**Pull request merged**